### DISCUSSION TITLE
How to tell if slot is filled

### DISCUSSION THREAD

**Discussion by alexgagnon** [lit/lit#2312] [Q&A] [answered]
Hi all, first time using Lit. I'm trying to return a different HTML structure depending on if a slot has been filled, but whenever I try to query how many nodes are in the slot, it's always 0. Here's the class:

```ts
// imports ...

export class Alert extends LitElement {
  @property({type: String})
  content: string | undefined = undefined;

  @queryAssignedNodes('heading', true)
  _headingNodes: NodeListOf<HTMLElement>;

  render() {
    return this._headingNodes.length === 0
      ? html`<div><slot><p>${this.content}</p></slot></div>`
      : html`<section><slot name='heading'></slot><slot><p>${this.content}</p></slot></section>`
    }
  }
}
```

and called like so:

```html
<wc-alert>This is some content</wc-alert>
<wc-alert><h2 slot='heading'>Alert!</h2>This is some more content</wc-alert>
```

It seems to always have length === 0, but what I'd like is that when the 'heading' slot is filled, the `<section />` variant is returned. Thanks!

**Comment by aomarks** [comment, marked as answer]
The problem you're hitting in your current implementation is that the slot you're checking with `this._headingNodes.length` never exists, so the count of slotted children will always be 0 (because the `queryAssignedNodes` decorator defaults to 0 if the slot doesn't exist). The slot doesn't exist on first render because no child DOM exists yet on first render. And actually even if something triggered a subsequent render, the count will still always be 0, because only the default slot could have been rendered, and never the `heading` slot you're looking for.

Here's a [playground](https://lit.dev/playground/#project=W3sibmFtZSI6IndjLWFsZXJ0LnRzIiwiY29udGVudCI6ImltcG9ydCB7aHRtbCwgY3NzLCBMaXRFbGVtZW50LCBub3RoaW5nfSBmcm9tICdsaXQnO1xuaW1wb3J0IHtjdXN0b21FbGVtZW50LCBxdWVyeUFzc2lnbmVkTm9kZXMsIHN0YXRlfSBmcm9tICdsaXQvZGVjb3JhdG9ycy5qcyc7XG5pbXBvcnQge2NsYXNzTWFwfSBmcm9tICdsaXQvZGlyZWN0aXZlcy9jbGFzcy1tYXAuanMnO1xuXG5AY3VzdG9tRWxlbWVudCgnd2MtYWxlcnQnKVxuZXhwb3J0IGNsYXNzIFdjQWxlcnQgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgc3RhdGljIHN0eWxlcyA9IGNzc2BcbiAgICBkaXYge1xuICAgICAgY29sb3I6IGJsdWU7XG4gICAgfVxuICAgIHNlY3Rpb24ge1xuICAgICAgY29sb3I6IHJlZDtcbiAgICB9XG4gIGA7XG4gIFxuICBAcXVlcnlBc3NpZ25lZE5vZGVzKCdoZWFkaW5nJywgdHJ1ZSlcbiAgcHJpdmF0ZSBfaGVhZGluZ05vZGVzOiBOb2RlTGlzdE9mPEhUTUxFbGVtZW50PjtcbiAgXG4gIEBzdGF0ZSgpXG4gIHByaXZhdGUgX2hhc0hlYWRpbmcgPSBmYWxzZTtcbiAgXG4gIHJlbmRlcigpIHtcbiAgICByZXR1cm4gdGhpcy5faGFzSGVhZGluZyA_IGh0bWxgXG4gICAgICA8ZGl2PlxuICAgICAgICA8c2xvdCBuYW1lPVwiaGVhZGluZ1wiIEBzbG90Y2hhbmdlPSR7dGhpcy5fb25IZWFkaW5nU2xvdENoYW5nZX0-PC9zbG90PlxuICAgICAgICA8cD5cbiAgICAgICAgICA8c2xvdD48L3Nsb3Q-XG4gICAgICAgIDwvcD5cbiAgICAgIDwvZGl2PlxuICAgIGAgOiBodG1sYFxuICAgICAgPHNlY3Rpb24-XG4gICAgICAgIDxzbG90IG5hbWU9XCJoZWFkaW5nXCIgQHNsb3RjaGFuZ2U9JHt0aGlzLl9vbkhlYWRpbmdTbG90Q2hhbmdlfT48L3Nsb3Q-XG4gICAgICAgIDxwPlxuICAgICAgICAgIDxzbG90Pjwvc2xvdD5cbiAgICAgICAgPC9wPlxuICAgICAgPC9zZWN0aW9uPiAgICAgICAgXG4gICAgYDtcbiAgfVxuICBcbiAgcHJpdmF0ZSBfb25IZWFkaW5nU2xvdENoYW5nZSgpIHtcbiAgICB0aGlzLl9oYXNIZWFkaW5nID0gdGhpcy5faGVhZGluZ05vZGVzLmxlbmd0aCA-IDA7XG4gIH1cbn1cbiJ9LHsibmFtZSI6ImluZGV4Lmh0bWwiLCJjb250ZW50IjoiPCFET0NUWVBFIGh0bWw-XG48aGVhZD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi93Yy1hbGVydC5qc1wiPjwvc2NyaXB0PlxuPC9oZWFkPlxuPGJvZHk-XG5cbiAgPHdjLWFsZXJ0PlxuICAgIFRoaXMgaXMgc29tZSBjb250ZW50XG4gIDwvd2MtYWxlcnQ-XG4gIFxuICA8d2MtYWxlcnQ-XG4gICAgPGgyIHNsb3Q9J2hlYWRpbmcnPkFsZXJ0ITwvaDI-XG4gICAgVGhpcyBpcyBzb21lIGNvbnRlbnRcbiAgPC93Yy1hbGVydD5cbiAgXG48L2JvZHk-XG4ifSx7Im5hbWUiOiJwYWNrYWdlLmpzb24iLCJjb250ZW50Ijoie1xuICBcImRlcGVuZGVuY2llc1wiOiB7XG4gICAgXCJsaXRcIjogXCJeMi4wLjBcIixcbiAgICBcIkBsaXQvcmVhY3RpdmUtZWxlbWVudFwiOiBcIl4xLjAuMFwiLFxuICAgIFwibGl0LWVsZW1lbnRcIjogXCJeMy4wLjBcIixcbiAgICBcImxpdC1odG1sXCI6IFwiXjIuMC4wXCJcbiAgfVxufSIsImhpZGRlbiI6dHJ1ZX1d) which includes a `heading` slot even in the first case, because otherwise there is no way to check if anything was slotted into it, and adds `slotchange` listeners to trigger a re-render when the layout needs to change.

If possible though, I'd also suggest modifying this template so that it doesn't require switching the parent element type between `<div>` and `<section>`, because doing that requires adding/removing `<slot>` elements and those event listeners whenever the layout changes, and is more verbose. Here's another [playground](https://lit.dev/playground/#project=W3sibmFtZSI6IndjLWFsZXJ0LnRzIiwiY29udGVudCI6ImltcG9ydCB7aHRtbCwgY3NzLCBMaXRFbGVtZW50LCBub3RoaW5nfSBmcm9tICdsaXQnO1xuaW1wb3J0IHtjdXN0b21FbGVtZW50LCBxdWVyeUFzc2lnbmVkTm9kZXMsIHN0YXRlfSBmcm9tICdsaXQvZGVjb3JhdG9ycy5qcyc7XG5cbkBjdXN0b21FbGVtZW50KCd3Yy1hbGVydCcpXG5leHBvcnQgY2xhc3MgV2NBbGVydCBleHRlbmRzIExpdEVsZW1lbnQge1xuICBzdGF0aWMgc3R5bGVzID0gY3NzYFxuICAgIDpob3N0ID4gZGl2IHtcbiAgICAgIGNvbG9yOiBibHVlO1xuICAgIH1cbiAgICA6aG9zdCA-IGRpdltoZWFkZWRdIHtcbiAgICAgIGNvbG9yOiByZWQ7XG4gICAgfVxuICBgO1xuICBcbiAgQHF1ZXJ5QXNzaWduZWROb2RlcygnaGVhZGluZycsIHRydWUpXG4gIHByaXZhdGUgX2hlYWRpbmdOb2RlczogTm9kZUxpc3RPZjxIVE1MRWxlbWVudD47XG4gIFxuICBAc3RhdGUoKVxuICBwcml2YXRlIF9oYXNIZWFkaW5nID0gZmFsc2U7XG4gIFxuICByZW5kZXIoKSB7XG4gICAgcmV0dXJuIGh0bWxgXG4gICAgICA8ZGl2ID9oZWFkZWQ9JHt0aGlzLl9oYXNIZWFkaW5nfT5cbiAgICAgICAgPHNsb3QgbmFtZT1cImhlYWRpbmdcIiBAc2xvdGNoYW5nZT0ke3RoaXMuX29uSGVhZGluZ1Nsb3RDaGFuZ2V9Pjwvc2xvdD5cbiAgICAgICAgPHA-XG4gICAgICAgICAgPHNsb3Q-PC9zbG90PlxuICAgICAgICA8L3A-XG4gICAgICA8L2Rpdj5cbiAgICBgO1xuICB9XG4gIFxuICBwcml2YXRlIF9vbkhlYWRpbmdTbG90Q2hhbmdlKCkge1xuICAgIHRoaXMuX2hhc0hlYWRpbmcgPSB0aGlzLl9oZWFkaW5nTm9kZXMubGVuZ3RoID4gMDtcbiAgfVxufVxuIn0seyJuYW1lIjoiaW5kZXguaHRtbCIsImNvbnRlbnQiOiI8IURPQ1RZUEUgaHRtbD5cbjxoZWFkPlxuICA8c2NyaXB0IHR5cGU9XCJtb2R1bGVcIiBzcmM9XCIuL3djLWFsZXJ0LmpzXCI-PC9zY3JpcHQ-XG48L2hlYWQ-XG48Ym9keT5cblxuICA8d2MtYWxlcnQ-XG4gICAgVGhpcyBpcyBzb21lIGNvbnRlbnRcbiAgPC93Yy1hbGVydD5cbiAgXG4gIDx3Yy1hbGVydD5cbiAgICA8aDIgc2xvdD0naGVhZGluZyc-QWxlcnQhPC9oMj5cbiAgICBUaGlzIGlzIHNvbWUgY29udGVudFxuICA8L3djLWFsZXJ0PlxuICBcbjwvYm9keT5cbiJ9LHsibmFtZSI6InBhY2thZ2UuanNvbiIsImNvbnRlbnQiOiJ7XG4gIFwiZGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcImxpdFwiOiBcIl4yLjAuMFwiLFxuICAgIFwiQGxpdC9yZWFjdGl2ZS1lbGVtZW50XCI6IFwiXjEuMC4wXCIsXG4gICAgXCJsaXQtZWxlbWVudFwiOiBcIl4zLjAuMFwiLFxuICAgIFwibGl0LWh0bWxcIjogXCJeMi4wLjBcIlxuICB9XG59IiwiaGlkZGVuIjp0cnVlfV0) which toggles an attribute when the layout changes, instead of replacing the entire template. This should be more efficient and fewer bytes than the first example.

If the reason you wanted to switch between `<div>` and `<section>` was for semantic ARIA reasons, note that `<div role="region" aria-label="My section">` and `<section aria-label="My section">` are equivalent, so you could conditionally add those attributes if needed. (`<div>` and `<section>` without a label are already equivalent, since neither will have a role).

Also check out https://github.com/WICG/webcomponents/issues/936 for a proposal to add something like `:hasSlotted` which would allow you to write an equivalent solution to the second example purely in CSS. Give it a thumbs up if you agree!

**Reply by alexgagnon** [reply]
Thanks @aomarks! So if I understand correctly, on first pass there is no child DOM element created yet so it would register as 0, and then there's nothing to attach an event listener to to update it once the child DOM element would have been created?

And good to know about them behaving the same, I was following our orgs css/style guide but if I can push for just divs (with extra a11y labelling as required) that works.
